File: _1_data_structures_go/_3_stack_queue/17-next-greater-element4.go
// https://leetcode.com/problems/next-greater-element-iv/
// https://leetcode.com/problems/next-greater-element-iv/solutions/2799346/monotonic-stack-min-heap/

package main

import (
	"container/heap"
	"fmt"
)

/**************** MIN HEAP ****************/

type Pair struct {
	val int
	idx int
}

type MinHeap []Pair

func (h MinHeap) Len() int           { return len(h) }
func (h MinHeap) Less(i, j int) bool { return h[i].val < h[j].val }
func (h MinHeap) Swap(i, j int)      { h[i], h[j] = h[j], h[i] }

func (h *MinHeap) Push(x interface{}) {
	*h = append(*h, x.(Pair))
}

func (h *MinHeap) Pop() interface{} {
	old := *h
	n := len(old)
	item := old[n-1]
	*h = old[:n-1]
	return item
}

/**************** SOLUTION ****************/

func secondGreaterElement(nums []int) []int {

	n := len(nums)
	gE := make([]int, n)

	for i := range gE {
		gE[i] = -1
	}

	stack := []int{}
	pq := &MinHeap{}
	heap.Init(pq)

	for i := 0; i < n; i++ {

		// resolve second greater
		for pq.Len() > 0 && (*pq)[0].val < nums[i] {
			top := heap.Pop(pq).(Pair)
			gE[top.idx] = nums[i]
		}

		// move first greater candidates
		for len(stack) > 0 && nums[stack[len(stack)-1]] < nums[i] {
			idx := stack[len(stack)-1]
			stack = stack[:len(stack)-1]
			heap.Push(pq, Pair{nums[idx], idx})
		}

		stack = append(stack, i)
	}

	return gE
}

/**************** TEST ****************/

func main() {

	nums := []int{2, 4, 0, 9, 6}

	res := secondGreaterElement(nums)

	fmt.Println(res)
}

File: _1_data_structures/_1_data_structures_cpp/_3_stack_and_queue/_17_next_greater_element4.cpp
// https://leetcode.com/problems/next-greater-element-iv/
// https://leetcode.com/problems/next-greater-element-iv/solutions/2799346/monotonic-stack-min-heap/


class Solution {
public:
    vector<int> secondGreaterElement(vector<int>& nums) {
        int n = nums.size();
        vector<int> gE(n, -1);
        stack<int> st;
        priority_queue<pair<int, int>, vector<pair<int, int>>, greater<pair<int, int>>> pq;
        
        for(int i = 0; i < n; i++) {
            while(!pq.empty() && pq.top().first < nums[i]) {
                gE[pq.top().second] = nums[i];
                pq.pop();
            }
            while(!st.empty() && nums[st.top()] < nums[i]) {
                pq.push(make_pair(nums[st.top()], st.top()));
                st.pop();
            }
            st.push(i);
        }
        
        return gE;
    }
};